### PR TITLE
Revert revert of equivalence class hash calculation in scheduler

### PR DESCRIPTION
**Pull request by resouer**
**What this PR does / why we need it**:

NOTE: This is a revert revert of https://github.com/kubernetes/kubernetes/pull/58555

But since the original PR has been changed, I have to copy the original changes and resend this new PR. See: https://github.com/kubernetes/kubernetes/pull/58555#issuecomment-364345972

And I kept @misterikkit 's change as the first commit (by co-author feature of github) in the history. 

We decide to do revert revert because #58989 has been fixed, which should help to improve the time consumed by integration test.

**But** we should still pay attention to integration tests to see if there's frequent timeout happen.

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by thockin** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/60753)

**Comment by resouer** [comment]
The root cause is found.

1. After changing the hash calculation method, `TestVolumeBinding` will fail (local), in CI, it actually took a long time to check failure and cause timeout.
```


--- FAIL: TestVolumeBinding (198.46s)
	volume_binding_test.go:208: Failed to schedule Pod "pod-i-canbind": timed out waiting for the condition
	volume_binding_test.go:204: Pod "pod-i-cannotbind" was not unschedulable: timed out waiting for the condition
	volume_binding_test.go:208: Failed to schedule Pod "pod-i-pvc-prebound": timed out waiting for the condition
	volume_binding_test.go:208: Failed to schedule Pod "pod-i-pv-prebound": timed out waiting for the condition
	volume_binding_test.go:208: Failed to schedule Pod "pod-w-canbind": timed out waiting for the condition
	volume_binding_test.go:595: PVC volume-scheduling/pvc-w-canbind phase not Bound, got Pending
	volume_binding_test.go:606: PV pv-w-canbind phase not Bound, got Available
	volume_binding_test.go:204: Pod "pod-w-cannotbind" was not unschedulable: timed out waiting for the condition
```
2. Also, `TestVolumeBindingStress` will also take a long time and timeout (just local), not sure if it's enabled in CI.

**Comment by bsalamat** [comment]
Thanks @resouer for investigating this. Please check that eCache is updated/invalidated properly on volume events. That could be the cause of the failures.

**Comment by msau42** [comment]
/assign

The volume integration tests don't cover equivalence cache invalidation scenarios yet. So if these volume tests are failing, then there's a more fundamental issue with the change in the hashing mechanism, perhaps causing multiple pods to look equivalent when they should not be.

**Comment by msau42** [comment]
If you want to turn up the log level to 5, and paste the logs somewhere from a test run, I can try to help triage.

**Comment by resouer** [comment]
@msau42 Thanks a lot! 

While I've raised the log level locally and found that `predicate()` of `CheckVolumeBinding` actually stuck at this line and timeout:
```go
	unboundSatisfied, boundSatisfied, err := c.binder.Binder.FindPodVolumes(pod, node.Name)
```
It is a real predicate process since ecache info is not found for that pod, which is expected.

**Comment by msau42** [comment]
Hm, I wonder if there is a deadlock somewhere?  Are you able to get the goroutine dump?

**Comment by resouer** [comment]
@msau42 @bsalamat OK, here is the bug, it's straightforward ... 

We introduced a global `ecache.Lock() ` here:

https://github.com/kubernetes/kubernetes/blob/71b40cbce509960cb020822c85e05b379c37d722/pkg/scheduler/core/generic_scheduler.go#L474-L480

But only did `ecache.unlock()` here:

https://github.com/kubernetes/kubernetes/blob/71b40cbce509960cb020822c85e05b379c37d722/pkg/scheduler/core/generic_scheduler.go#L502-L504

While when `VolumeBindingChecker` fail for the first time, it directly returns with error here: 

https://github.com/kubernetes/kubernetes/blob/71b40cbce509960cb020822c85e05b379c37d722/pkg/scheduler/core/generic_scheduler.go#L482-L487

So next iterate, when the pod been checked again (and will become schedulable), it deadlock.

I will fix this in following commit.

Also, I have a plan in mind to move "do this pod have equivalence class" check out from `PredicateWithECache`, so we don't have to call it everytime, would also help to make this logic clearer.

File: pkg/scheduler/core/generic_scheduler.go [hunk 1]
```diff
@@ -483,6 +483,10 @@ func podFitsOnNode(
 					// we need to execute predicate functions since equivalence cache does not work
 					fit, reasons, err = predicate(pod, metaToUse, nodeInfoToUse)
 					if err != nil {
+						// TODO(harry): We need to unlock ecache before return, but find a better way to do this.
```
Comment (bsalamat) [inline pkg/scheduler/core/generic_scheduler.go]: @resouer I think we should create an inline function whose body covers lines 474 to 508 (and I think the function should return `fit` and `reasons`). Then we should remove all these `ecache.Unlock()` calls and add `defer ecache.Unlock()` after each `ecache.Lock()`.

**Comment by bsalamat** [comment]
@resouer Great job finding the issue, BTW!  :smiley:

File: pkg/scheduler/core/generic_scheduler.go (hunk 1 above)
Comment (resouer) [reply pkg/scheduler/core/generic_scheduler.go]: Thanks! This is exactly what I am doing. While I added a explicit function named `doPredicate`, not sure what do you think it VS anonymous one?

File: pkg/scheduler/core/generic_scheduler.go [hunk 2]
```diff
@@ -469,38 +465,54 @@ func podFitsOnNode(
...
-					fit, reasons, invalid = ecache.PredicateWithECache(pod.GetName(), info.Node().GetName(), predicateKey, equivCacheInfo.hash, false)
-				}
-
-				if !eCacheAvailable || invalid {
-					// we need to execute predicate functions since equivalence cache does not work
-					fit, reasons, err = predicate(pod, metaToUse, nodeInfoToUse)
-					if err != nil {
-						return false, []algorithm.PredicateFailureReason{}, err
-					}
+				// Use a in-line function here so we can ensure ecache.Unlock() for the block in closure.
```
Comment (bsalamat) [inline pkg/scheduler/core/generic_scheduler.go]: nit: Please change the comment to: Use an in-line function to guarantee invocation of ecache.Unlock() when the inline function returns.

File: pkg/scheduler/core/generic_scheduler.go [hunk 3]
```diff
@@ -469,38 +465,54 @@ func podFitsOnNode(
...
+							result := predicateResults[predicateKey]
+							ecache.UpdateCachedPredicateItem(
+								pod.GetName(), info.Node().GetName(),
+								predicateKey, result.Fit, result.FailReasons, equivCacheInfo.hash, false)
 						}
-						result := predicateResults[predicateKey]
-						ecache.UpdateCachedPredicateItem(pod.GetName(), info.Node().GetName(), predicateKey, result.Fit, result.FailReasons, equivCacheInfo.hash, false)
 					}
-				}
+					return
```
Comment (bsalamat) [inline pkg/scheduler/core/generic_scheduler.go]: Please remove the "return"

**Review by bsalamat — COMMENTED** [review]
Looks good. Only a couple of nits.

**Comment by bsalamat** [comment]
/lgtm

Thanks, @resouer!

**Pull request merged**